apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java

```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
-                excluded.set(Boolean.TRUE);
-                return;
-            }
-            final Request req = transaction.getContext().getRequest();
-            if (transaction.isSampled() && tracer.getConfig(CoreConfiguration.class).isCaptureHeaders()) {
-                if (request.getCookies() != null) {
-                    for (Cookie cookie : request.getCookies()) {
-                        req.addCookie(cookie.getName(), cookie.getValue());
+                if (transaction == null) {
+                    // if the request is excluded, avoid matching all exclude patterns again on each filter invocation
+                    excluded.set(Boolean.TRUE);
+                    return;
+                }
+                final Request req = transaction.getContext().getRequest();
+                if (transaction.isSampled() && tracer.getConfig(CoreConfiguration.class).isCaptureHeaders()) {
+                    if (request.getCookies() != null) {
+                        for (Cookie cookie : request.getCookies()) {
+                            req.addCookie(cookie.getName(), cookie.getValue());
+                        }
+                    }
+                    final Enumeration<String> headerNames = request.getHeaderNames();
+                    if (headerNames != null) {
+                        while (headerNames.hasMoreElements()) {
+                            final String headerName = headerNames.nextElement();
+                            req.addHeader(headerName, request.getHeaders(headerName));
+                        }
                     }
                 }
-                final Enumeration<String> headerNames = request.getHeaderNames();
-                if (headerNames != null) {
-                    while (headerNames.hasMoreElements()) {
-                        final String headerName = headerNames.nextElement();
-                        req.addHeader(headerName, request.getHeaders(headerName));
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+
+                servletTransactionHelper.fillRequestContext(transaction, request.getProtocol(), request.getMethod(), request.isSecure(),
+                    request.getScheme(), request.getServerName(), request.getServerPort(), request.getRequestURI(), request.getQueryString(),
+                    request.getRemoteAddr(), request.getHeader("Content-Type"));
+            } else if (transaction == null) {
+                final AbstractSpan<?> parent = tracer.getActive();
+                if (parent != null) {
+                    final HttpServletRequest request = (HttpServletRequest) servletRequest;
+                    DispatcherType dispatcherType = request.getDispatcherType();
+                    boolean isSpannableDispatcherType = false;
+                    span = parent.createSpan()
+                        .withType(SPAN_TYPE)
+                        .withSubtype(SPAN_SUBTYPE);
```

Comment:
Do not create a span unless you know we need one. Currently these spans can leak. You must always end a span you created, like you must always deactivate a span you activated (which you did).
In rare cases (this is not one), you may create a span and then `requestDiscarding` it, but you still must end it.